internal/dataplane/parser/translators/ingress_atc.go

```diff
@@ -0,0 +1,243 @@
+	ingressHost := m.ingressHost
+	if strings.Contains(ingressHost, "*") {
+		// '_' is not allowed in host, so we use '_' to replace '*' since '*' is not allowed in Kong.
+		ingressHost = strings.ReplaceAll(ingressHost, "*", "_")
+	}
+
+	routeName := fmt.Sprintf("%s.%s.%s.%s.%s", m.parentIngress.GetNamespace(), m.parentIngress.GetName(), m.serviceName, ingressHost, m.servicePort.CanonicalString())
+	route := &kongstate.Route{
+		Ingress: util.K8sObjectInfo{
+			Namespace:   m.parentIngress.GetNamespace(),
+			Name:        m.parentIngress.GetName(),
+			Annotations: m.parentIngress.GetAnnotations(),
+		},
+		Route: kong.Route{
+			Name:              kong.String(routeName),
+			StripPath:         kong.Bool(false),
+			PreserveHost:      kong.Bool(true),
+			RequestBuffering:  kong.Bool(true),
+			ResponseBuffering: kong.Bool(true),
+			Tags:              m.ingressTags,
+		},
+		ExpressionRoutes: true,
+	}
+
+	ingressAnnotations := m.parentIngress.GetAnnotations()
+
+	routeMatcher := atc.And()
+	// translate hosts.
+	hosts := []string{}
+	if m.ingressHost != "" {
+		hosts = append(hosts, m.ingressHost)
+	}
+	hostAliases, _ := annotations.ExtractHostAliases(ingressAnnotations)
+	hosts = append(hosts, hostAliases...)
+	if len(hosts) > 0 {
+		hostMatcher := hostMatcherFromIngressHosts(hosts)
+		routeMatcher.And(hostMatcher)
+	}
+
+	// translate paths.
+	pathMatchers := make([]atc.Matcher, 0, len(m.paths))
+	pathRegexPrefix := annotations.ExtractRegexPrefix(ingressAnnotations)
+	if pathRegexPrefix == "" {
+		pathRegexPrefix = ControllerPathRegexPrefix
+	}
+	for _, path := range m.paths {
+		pathMatchers = append(pathMatchers, pathMatcherFromIngressPath(path, pathRegexPrefix))
+	}
+	routeMatcher.And(atc.Or(pathMatchers...))
+
+	// translate protocols.
+	protocols := []string{"http", "https"}
+	annonationProtocols := annotations.ExtractProtocolNames(ingressAnnotations)
+	if len(annonationProtocols) > 0 {
+		protocols = annonationProtocols
+	}
+	protocolMatcher := protocolMatcherFromProtocols(protocols)
+	routeMatcher.And(protocolMatcher)
+
+	// translate headers.
+	headers, exist := annotations.ExtractHeaders(ingressAnnotations)
+	if len(headers) > 0 && exist {
+		headerMatcher := headerMatcherFromHeaders(headers)
+		routeMatcher.And(headerMatcher)
+	}
+
+	// translate methods.
+	methods := annotations.ExtractMethods(ingressAnnotations)
+	if len(methods) > 0 {
+		methodMatcher := methodMatcherFromMethods(methods)
+		routeMatcher.And(methodMatcher)
+	}
+
+	// translate SNIs.
+	snis, exist := annotations.ExtractSNIs(ingressAnnotations)
+	if exist && len(snis) > 0 {
+		sniMatcher := sniMatcherFromSNIs(snis)
+		routeMatcher.And(sniMatcher)
+	}
+
+	atc.ApplyExpression(&route.Route, routeMatcher, NormalIngressExpressionPriority)
+	return route
+}
+
+// hostMatcherFromIngressHosts translates hosts in IngressHost format to ATC matcher that matches any of them.
+func hostMatcherFromIngressHosts(hosts []string) atc.Matcher {
```

Comment:
IMHO all of the `*MatcherFrom*` funcs deserve a separate unit test looking at their quite complex logic. Would you mind adding those?

Reply:
Added unit tests.